libminifi/include/core/Property.h

```diff
@@ -129,41 +133,32 @@ class Property {
 
   template<typename T = std::string>
   void setValue(const T &value) {
-    PropertyValue vn = default_value_;
-    vn = value;
-    if (validator_) {
-      vn.setValidator(validator_);
-      ValidationResult result = validator_->validate(name_, vn.getValue());
-      if (!result.valid()) {
-        // throw some exception?
-      }
-    } else {
-      vn.setValidator(core::StandardValidators::VALID);
-    }
     if (!is_collection_) {
       values_.clear();
-      values_.push_back(vn);
+      values_.push_back(default_value_);
     } else {
-      values_.push_back(vn);
+      values_.push_back(default_value_);
     }
+    PropertyValue& vn = values_.back();
+    vn.setValidator(validator_ ? validator_ : core::StandardValidators::VALID);
+    vn = value;
+    ValidationResult result = vn.validate(name_);
+    if(!result.valid())
+      throw utils::InvalidValueException(name_ + " value validation failed");
   }
 
-  void setValue(PropertyValue &vn) {
-    if (validator_) {
-      vn.setValidator(validator_);
-      ValidationResult result = validator_->validate(name_, vn.getValue());
-      if (!result.valid()) {
-        // throw some exception?
-      }
-    } else {
-      vn.setValidator(core::StandardValidators::VALID);
-    }
+  void setValue(PropertyValue &newValue) {
     if (!is_collection_) {
       values_.clear();
-      values_.push_back(vn);
+      values_.push_back(newValue);
     } else {
-      values_.push_back(vn);
+      values_.push_back(newValue);
     }
+    PropertyValue& vn = values_.back();
+    vn.setValidator(validator_ ? validator_ : core::StandardValidators::VALID);
```

Comment:
it sets the validator in the PropertyValue, if we didn't the incoming `newValue`'s validator would remain in the copy, yes we could also just clear the validator as most places interpret the absence of a validator as being valid, but I find it unappealing, imo we should always have a non-empty validator instance

Reply:
I think it would be cleaner to have either one or the other. We could set the validator true by default and never expect it missing, OR handle the no-validator case as a valid option (both literally and implementationwise).

Reply:
yes, replacing `std::shared_ptr<PropertyValidator>` with a suitable wrapper that guarantees a `PropertyValidator` instance is viable, but since most API functions expect/return a `shared_ptr` I would wait for after `0.8.0` to make the change

Reply:
propagated `gsl::not_null` all around, had to remove `StandardValidator::VALID` as multiple statically initialized `Property`s use it